x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyMaintenanceService.java

```diff
@@ -0,0 +1,173 @@
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.LocalNodeMasterListener;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.component.LifecycleListener;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.common.xcontent.ObjectPath;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.enrich.EnrichPolicy;
+
+public class EnrichPolicyMaintenanceService implements LocalNodeMasterListener {
+
+    private static final Logger logger = LogManager.getLogger(EnrichPolicyMaintenanceService.class);
+
+    private static final String MAPPING_POLICY_FIELD_PATH = MapperService.SINGLE_MAPPING_NAME + "._meta." +
+        EnrichPolicyRunner.ENRICH_POLICY_FIELD_NAME;
+    private static final IndicesOptions IGNORE_UNAVAILABLE = IndicesOptions.fromOptions(true, false, false, false);
+
+    private final Settings settings;
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final EnrichPolicyLocks enrichPolicyLocks;
+
+    private volatile Scheduler.Cancellable cancellable;
+
+    EnrichPolicyMaintenanceService(Settings settings, Client client, ClusterService clusterService, ThreadPool threadPool,
+                                   EnrichPolicyLocks enrichPolicyLocks) {
+        this.settings = settings;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.enrichPolicyLocks = enrichPolicyLocks;
+    }
+
+    void initialize() {
+        clusterService.addLocalNodeMasterListener(this);
+    }
+
+    @Override
+    public void onMaster() {
+        if (cancellable == null || cancellable.isCancelled()) {
+            scheduleNext();
+            clusterService.addLifecycleListener(new LifecycleListener() {
+                @Override
+                public void beforeStop() {
+                    offMaster();
+                }
+            });
+        }
+    }
+
+    @Override
+    public void offMaster() {
+        if (cancellable != null && cancellable.isCancelled() == false) {
+            cancellable.cancel();
+        }
+    }
+
+    @Override
+    public String executorName() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    private void scheduleNext() {
+        try {
+            TimeValue waitTime = EnrichPlugin.ENRICH_CLEANUP_PERIOD.get(settings);
+            cancellable = threadPool.schedule(this::execute, waitTime, ThreadPool.Names.GENERIC);
```

Comment:
I think we need to have a field that indices that this node is still active master and check that here before scheduling. Because I think now if `offMaster()` is invoked, there is still a chance that `scheduleNext()` gets invoked that sets a new value to the `cancellable` field?